File: go/spore-core/observability/observability_fixture_replay_test.go
package observability

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/squirrelsoft-dev/spore-core/go/spore-core/guideregistry"
)

// fixtureCase mirrors the shared JSON shape in
// fixtures/observability/session_metrics_basic.json so the same fixture
// deserialises across all language implementations (Rust reference at
// `rust/crates/spore-core/src/observability.rs::FixtureCase`).
type fixtureCase struct {
	SessionID string        `json:"session_id"`
	Turns     []fixtureTurn `json:"turns"`
	Outcome   string        `json:"outcome"`
	Expected  fixtureExpect `json:"expected"`
}

type fixtureTurn struct {
	SpanID string `json:"span_id"`
	Turn   uint32 `json:"turn"`
	Input  uint32 `json:"input"`
	Output uint32 `json:"output"`
}

type fixtureExpect struct {
	TotalTurns        uint32 `json:"total_turns"`
	TotalInputTokens  uint32 `json:"total_input_tokens"`
	TotalOutputTokens uint32 `json:"total_output_tokens"`
}

// TestObservabilityFixtureReplay loads the shared session_metrics_basic.json
// fixture, replays each turn through the in-memory provider, and asserts the
// aggregated SessionMetrics match the fixture's expected values.
func TestObservabilityFixtureReplay(t *testing.T) {
	wd, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	// go/spore-core/observability → ../../../fixtures/observability/...
	path := filepath.Join(wd, "..", "..", "..", "fixtures", "observability", "session_metrics_basic.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fc fixtureCase
	if err := json.Unmarshal(raw, &fc); err != nil {
		t.Fatalf("unmarshal fixture: %v", err)
	}

	obs := NewInMemoryObservabilityProvider()
	for _, ft := range fc.Turns {
		obs.EmitTurn(turnSpan(fc.SessionID, ft.SpanID, ft.Turn, ft.Input, ft.Output))
	}
	var outcome guideregistry.SessionOutcome
	switch fc.Outcome {
	case "success":
		outcome = guideregistry.NewOutcomeSuccess()
	case "partial":
		outcome = guideregistry.NewOutcomePartial()
	default:
		outcome = guideregistry.NewOutcomeFailure(fc.Outcome)
	}
	obs.SetSessionOutcome(sid(fc.SessionID), outcome)

	m, err := obs.GetSessionMetrics(context.Background(), sid(fc.SessionID))
	if err != nil || m == nil {
		t.Fatalf("metrics: err=%v m=%v", err, m)
	}
	if m.TotalTurns != fc.Expected.TotalTurns {
		t.Fatalf("total_turns=%d want %d", m.TotalTurns, fc.Expected.TotalTurns)
	}
	if m.TotalInputTokens != fc.Expected.TotalInputTokens {
		t.Fatalf("total_input_tokens=%d want %d", m.TotalInputTokens, fc.Expected.TotalInputTokens)
	}
	if m.TotalOutputTokens != fc.Expected.TotalOutputTokens {
		t.Fatalf("total_output_tokens=%d want %d", m.TotalOutputTokens, fc.Expected.TotalOutputTokens)
	}
}

File: typescript/packages/core/tests/observability-fixture-replay.test.ts
/**
 * Fixture-replay tests for the canonical ObservabilityProvider
 * (spore-core issue #12).
 *
 * Loads `fixtures/observability/session_metrics_basic.json` and asserts the
 * aggregated totals exactly match the Rust, Python, and Go suites.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { memory, observability, SessionId, TaskId, type guideRegistry } from "../src/index.js";

const { InMemoryObservabilityProvider, SpanId } = observability;
const { Timestamp } = memory;

interface FixtureTurn {
  span_id: string;
  turn: number;
  input: number;
  output: number;
}

interface FixtureCase {
  session_id: string;
  turns: FixtureTurn[];
  outcome: string;
  expected: {
    total_turns: number;
    total_input_tokens: number;
    total_output_tokens: number;
  };
}

const here = dirname(fileURLToPath(import.meta.url));
const fixturePath = resolve(here, "../../../../fixtures/observability/session_metrics_basic.json");

function outcomeOf(raw: string): guideRegistry.SessionOutcome {
  switch (raw) {
    case "success":
      return { kind: "success" };
    case "partial":
      return { kind: "partial" };
    default:
      return { kind: "failure", reason: raw };
  }
}

describe("Observability fixture replay", () => {
  it("session_metrics_basic.json aggregates match expected totals", async () => {
    const raw = readFileSync(fixturePath, "utf8");
    const fixture = JSON.parse(raw) as FixtureCase;

    const obs = new InMemoryObservabilityProvider();
    const sessionId = SessionId.of(fixture.session_id);
    const taskId = TaskId.of("fixture_task");

    for (const t of fixture.turns) {
      obs.emitTurn({
        base: {
          span_id: SpanId.of(t.span_id),
          parent_span_id: null,
          session_id: sessionId,
          task_id: taskId,
          kind: "turn",
          started_at: Timestamp.of("2026-05-16T00:00:00Z"),
          ended_at: Timestamp.of("2026-05-16T00:00:00Z"),
          duration_ms: 0,
          status: { kind: "ok" },
        },
        turn_number: t.turn,
        input_tokens: t.input,
        output_tokens: t.output,
        cache_read_tokens: null,
        cache_write_tokens: null,
        cost_usd: 0,
        stop_reason: "end_turn",
        tool_calls_requested: 0,
      });
    }
    obs.setSessionOutcome(sessionId, outcomeOf(fixture.outcome));

    const metrics = await obs.getSessionMetrics(sessionId);
    expect(metrics).toBeDefined();
    expect(metrics!.total_turns).toBe(fixture.expected.total_turns);
    expect(metrics!.total_input_tokens).toBe(fixture.expected.total_input_tokens);
    expect(metrics!.total_output_tokens).toBe(fixture.expected.total_output_tokens);
  });
});
